Implement extend_from_slice and insert_from_slice with memmove optimization

nipunn1313 opened the pull request:
Implement the performance optimized versions of insert_many and extend (from #28). These methods use memmove rather than a looped insert.

If we had function specialization, we could implement these without exposing new methods. Up to the maintainers whether we want to support these new methods.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/rust-smallvec/29)

<!-- Reviewable:end -->

Marwes commented:
Might want to call `extend_slice`, [copy_from_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.copy_from_slice) to mimic the `method on `Vec`

emilio commented on lib.rs:
```diff
@@ -315,6 +315,29 @@ impl<A: Array> SmallVec<A> {
 
+impl<A: Array> SmallVec<A> where A::Item: Copy {
+    pub fn insert_slice(&mut self, index: usize, slice: &[A::Item]) {
+        self.reserve(slice.len());
```
Hmm... I'm kind of sleepy right now, but I think here you should reserve `index + slice.len()`?

nipunn1313 replied:
I believe reserve takes in an "additional" amount required and ensures the capacity is large enough, so I think this is ok.

emilio replied:
Oh, you're right, nvm, I thought `reserve()` did the usual stuff, that is, ensuring capacity.

nipunn1313 commented:
I think "extend_from_slice" and "insert_from_slice" are also reasonable names. I think "copy_from_slice" is misleading because it's not overwriting existing contents.

Marwes commented:
[extend_from_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice) is also a method on `Vec` except it takes `&[T] where T: Clone`.

nipunn1313 commented:
That makes sense. I agree with that.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #28) made this pull request unmergeable. Please resolve the merge conflicts.

nipunn1313 commented:
I went ahead and rebased this on master. Here are the benchmarks on my local machine

```
Nipunns-MacBook-Pro:rust-smallvec nipunn$ cargo bench bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/bench-760919c65e248f9d

running 7 tests
test bench_extend            ... bench:         149 ns/iter (+/- 26)
test bench_extend_from_slice ... bench:          65 ns/iter (+/- 8)
test bench_insert            ... bench:       1,411 ns/iter (+/- 440)
test bench_insert_from_slice ... bench:         119 ns/iter (+/- 22)
test bench_insert_many       ... bench:         483 ns/iter (+/- 59)
test bench_push              ... bench:         425 ns/iter (+/- 217)
test bench_pushpop           ... bench:         351 ns/iter (+/- 78)

test result: ok. 0 passed; 0 failed; 0 ignored; 7 measured

     Running target/release/deps/smallvec-82287e02aa187b8d

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

Nipunns-MacBook-Pro:rust-smallvec nipunn$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
Nipunns-MacBook-Pro:rust-smallvec nipunn$ cargo bench bench
   Compiling smallvec v0.3.1 (file:///Users/nipunn/src/rust-smallvec)
    Finished release [optimized] target(s) in 2.54 secs
     Running target/release/bench-760919c65e248f9d

running 5 tests
test bench_extend      ... bench:         151 ns/iter (+/- 16)
test bench_insert      ... bench:       1,407 ns/iter (+/- 369)
test bench_insert_many ... bench:         480 ns/iter (+/- 75)
test bench_push        ... bench:         430 ns/iter (+/- 152)
test bench_pushpop     ... bench:         353 ns/iter (+/- 104)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured

     Running target/release/deps/smallvec-82287e02aa187b8d

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
```

Looks like there shouldn't be regressions, though we should be extra careful to make sure there aren't integer overflow / memory issues.

jdm commented:
@bors-servo: r+
Thanks!

bors-servo commented:
:pushpin: Commit aef2113 has been approved by `jdm`

<!-- @bors-servo r=jdm aef21138179cf7b8a3306429f0c06628a40da15e -->

bors-servo commented:
:zap: Test exempted - [status](https://travis-ci.org/servo/rust-smallvec/builds/195087508)

nox commented:
Why does that use `Copy` instead of `Clone` like the stdlib?

nipunn1313 commented:
unsafe ptr::copy will only work if the type is Copy.
If the type is clone, then we have to iterate over the elements and run the clone code.
You're right that it's a little bit different than the stdlib which is unfortunate.

If function specialization lands (https://github.com/rust-lang/rust/issues/31844), then this naming / perf tradeoff may become moot.

nox commented:
The stdlib uses `Clone` but manages to optimise it nonetheless without specialisation AFAIK. Just because you iterate and clone doesn't mean it has to be slow.

nox commented:
Ah no, nevermind: https://github.com/rust-lang/rust/blob/master/src/libcollections/vec.rs#L1691-L1703

nipunn1313 commented:
Interesting! That looks like it uses specialization. At the time we made
this PR, specialization wasn't available.
--Nipunn

On Thu, Jan 26, 2017 at 2:32 AM Anthony Ramine <notifications@github.com>
wrote:

> Ah no, nevermind:
> https://github.com/rust-lang/rust/blob/master/src/libcollections/vec.rs#L1691-L1703
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/rust-smallvec/pull/29#issuecomment-275356436>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPXow0e9dYXtbdo2Rm412-xJaPeOGIkks5rWHYxgaJpZM4JrQG6>
> .
>
